### PR TITLE
Issue 3496 - Select media from wp media library - Part One

### PR DESCRIPTION
**Pull request by SergioEstevao**
A first step PR to address #3496. I split this fix in multiple PR in order to avoid a complex/lengthy review. 

This PR only allows insertion of site media that is already present on the device. It still doesn't sync with the server all media for the site.

To test it: 

 * Create a new post and insert some images/videos to it using the device media option.
 * After the insertion is done click on insert media again but this time select the site media. 
 * You should see the media that you previously inserted on the post for that blog.
 * You could also go to the feature image option on settings and use site media to select a previous inserted image on the post.

You may see some empty media items on the media picker, this is because we had a bug on the previous version of the app where the media items images where not being saved properly when a app update happen. Just ignore them or if you want  to start clean just delete the app from the simulator or device.

Needs Review: @bummytime

### CODE REVIEW

**Comment by bummytime** [comment]
@SergioEstevao Just so I understand, in Site Media, I should only see the images that I already uploaded from this device? (e.g. I should not be able to see any other media...)

**Comment by bummytime** [comment]
Is "Device Media" and "Site Media" something our users will understand clearly?

scaled screen shot 2015-06-11 at 11 24 11 am

/cc @meremagee

File: WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+-(void)loadDataWithSuccess:(WPMediaChangesBlock)successBlock failure:(WPMediaFailureBlock)failureBlock
+{
+    NSString *entityName = NSStringFromClass([Media class]);
+    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:entityName];
+    request.predicate = [[self class] predicateForFilter:self.filter];
+    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"creationDate" ascending:YES];
+    request.sortDescriptors = @[sortDescriptor];
+    NSError *error;
+    self.media = [[[ContextManager sharedInstance] mainContext] executeFetchRequest:request error:&error];
+    if (self.media == nil && error){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+-(id<WPMediaAsset>)mediaAtIndex:(NSInteger)index
+{
+    return self.media[index];
+}
+
++ (NSPredicate *)predicateForFilter:(WPMediaType)filter
+{
+    NSPredicate *predicate;
+    switch (filter){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        case WPMediaTypeImage:{
+            predicate = [NSPredicate predicateWithFormat:@"mediaTypeString = %@  && mediaID != 0", @"image"];
+        }break;
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m]: Not spelled out in the style guide...but should break be on a new line? I'll leave that up to you ;-)

File: WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    NSString *entityName = NSStringFromClass([Media class]);
+    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:entityName];
+    request.predicate = [MediaLibraryPickerDataSource predicateForFilter:self.filter];
+    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"creationDate" ascending:NO];
+    request.sortDescriptors = @[sortDescriptor];
+    NSError *error;
+    NSArray *mediaAssets = [[[ContextManager sharedInstance] mainContext] executeFetchRequest:request error:&error];
+    if (mediaAssets.count == 0)
+    {
+        if (completionHandler){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+- (WPMediaRequestID)imageWithSize:(CGSize)size completionHandler:(WPMediaImageBlock)completionHandler
+{
+    NSString *entityName = NSStringFromClass([Media class]);
+    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:entityName];
+    request.predicate = [MediaLibraryPickerDataSource predicateForFilter:self.filter];
+    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"creationDate" ascending:NO];
+    request.sortDescriptors = @[sortDescriptor];
+    NSError *error;
+    NSArray *mediaAssets = [[[ContextManager sharedInstance] mainContext] executeFetchRequest:request error:&error];
+    if (mediaAssets.count == 0)
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Media/MediaLibraryPickerDataSource.m]: Style nit: `{` on the same line as the `if` statement

File: WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m [hunk 6]
```diff
@@ -1082,25 +1124,35 @@ - (void)handleDateChange:(id)value
...
+    if (assets.count == 0 ){
         return;
     }
-    ALAsset *asset = [assets firstObject];
-    if (!asset.defaultRepresentation) {
-        [WPError showAlertWithTitle:NSLocalizedString(@"Image unavailable", @"The title for an alert that says the image the user selected isn't available.")
-                            message:NSLocalizedString(@"This Photo Stream image cannot be added to your WordPress. Try saving it to your Camera Roll before uploading.", @"User information explaining that the image is not available locally. This is normally related to share photo stream images.")  withSupportButton:NO];
-        return;
+    
+    if ([[assets firstObject] isKindOfClass:[ALAsset class]]){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m [hunk 7]
```diff
@@ -1082,25 +1124,35 @@ - (void)handleDateChange:(id)value
...
+        ALAsset *asset = [assets firstObject];
+        if (!asset.defaultRepresentation) {
+            [WPError showAlertWithTitle:NSLocalizedString(@"Image unavailable", @"The title for an alert that says the image the user selected isn't available.")
+                                message:NSLocalizedString(@"This Photo Stream image cannot be added to your WordPress. Try saving it to your Camera Roll before uploading.", @"User information explaining that the image is not available locally. This is normally related to share photo stream images.")  withSupportButton:NO];
+            return;
+        }
+        
+        self.isUploadingMedia = YES;
+        [self uploadFeatureImage:assets[0]];
+    } else if ([[assets firstObject] isKindOfClass:[Media class]]){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m [hunk 8]
```diff
@@ -1132,4 +1184,15 @@ - (BOOL)prefersStatusBarHidden
     return self.shouldHideStatusBar && !IS_IPAD;
 }
 
+#pragma mark - ActionSheet Delegate Methods
+
+
+- (void)actionSheet:(UIActionSheet *)actionSheet didDismissWithButtonIndex:(NSInteger)buttonIndex {
+    if (buttonIndex == actionSheet.firstOtherButtonIndex){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/PostSettingsViewController.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Post/WPLegacyEditPostViewController.m [hunk 9]
```diff
@@ -410,7 +453,7 @@ - (void)cancelEditing
                                          otherButtonTitles:NSLocalizedString(@"Update Draft", @"Button shown if there are unsaved changes and the author is trying to move away from an already published/saved post."), nil];
     }
 
-    actionSheet.tag = 201;
+    actionSheet.tag = WPLegacyEditPostViewControllerActionSheetCancelOptions;
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/WPLegacyEditPostViewController.m]: :+1:

File: WordPress/Classes/ViewRelated/Post/WPLegacyEditPostViewController.m [hunk 10]
```diff
@@ -924,7 +967,20 @@ - (void)prepareMediaProgressForNumberOfAssets:(NSUInteger)count
...
 }
 
-- (void) addMediaAssets:(NSArray *)assets
+- (void)addMediaAssets:(NSArray *)assets
+{
+    if (assets.count == 0) {
+        return;
+    }
+    id<WPMediaAsset> firstObject = [assets firstObject];
+    if ([firstObject isKindOfClass:[ALAsset class]]){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/WPLegacyEditPostViewController.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 11]
```diff
@@ -2147,29 +2210,37 @@ - (void)mediaPickerControllerDidCancel:(WPMediaPickerViewController *)picker
...
 }
 
-- (BOOL)mediaPickerController:(WPMediaPickerViewController *)picker shouldSelectAsset:(ALAsset *)asset
+- (BOOL)mediaPickerController:(WPMediaPickerViewController *)picker shouldSelectAsset:(id<WPMediaAsset>)mediaAsset
 {
-    NSString * assetType = [asset valueForProperty:ALAssetPropertyType];
-    
-    if (assetType == ALAssetTypeUnknown) {
-        return NO;
+    if ([mediaAsset isKindOfClass:[Media class]]){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Style nit: space between the `)` and `{`

File: WordPress/Classes/ViewRelated/Post/WPPostViewController.m [hunk 12]
```diff
@@ -2147,29 +2210,37 @@ - (void)mediaPickerControllerDidCancel:(WPMediaPickerViewController *)picker
...
-    if (!asset.defaultRepresentation) {
-        if (assetType == ALAssetTypePhoto) {
-            [WPError showAlertWithTitle:NSLocalizedString(@"Image unavailable", @"The title for an alert that says the image the user selected isn't available.")
-                                message:NSLocalizedString(@"This Photo Stream image cannot be added to your WordPress. Try saving it to your Camera Roll before uploading.", @"User information explaining that the video is not available locally. This is normally related to share photo stream images.")
-                      withSupportButton:NO];
-        } else {
-            [WPError showAlertWithTitle:NSLocalizedString(@"Video unavailable", @"The title for an alert that says the video the user selected isn't available.")
-                                message:NSLocalizedString(@"This Photo Stream video cannot be added to your WordPress. Try saving it to your Camera Roll before uploading.", @"User information explaining that the video is not available locally. This is normally related to share photo stream images.")
-                      withSupportButton:NO];
+    if ([mediaAsset isKindOfClass:[ALAsset class]]){
```
Comment (bummytime) [inline WordPress/Classes/ViewRelated/Post/WPPostViewController.m]: Style nit: space between the `)` and `{`

**Comment by bummytime** [comment]
@SergioEstevao On the iPad, it looks like the action sheet does not center properly when the device is rotated (after the action sheet is already displayed).

**Portrait -> Landscape:**

scaled screen shot 2015-06-11 at 11 45 11 am

**Landscape -> Portrait:**

scaled screen shot 2015-06-11 at 11 45 37 am

**Comment by bummytime** [comment]
Ok, @SergioEstevao that is my first pass on this PR. Lemme know if you have any questions.

**Comment by SergioEstevao** [comment]
@bummytime, @meremagee 
OK, I implemented the discussed changes and now you can add images while viewing the WP Media Library on the picker and I also joined both media sources in one picker.

You no longer see the action sheet, but you go directly to the picker and you see on the groups these options:

screen shot 2015-06-15 at 16 39 44

What do you think?

**Comment by bummytime** [comment]
@SergioEstevao I think this is looking better. It seems more straight forward for the user IMHO.

> What should be default group selected? The camera roll or the WP media library?

My vote would be for the Camera Roll. 

> Independently of the default I think the WP Media library and the Camera Roll should be the top options on the group selection.

:+1: 

@meremagee any thoughts on this?

**Comment by bummytime** [comment]
@SergioEstevao I forgot Mere was going to be gone this week. Let's just move forward with the label text you have....we can change it later if needed. I tested the new workflow out on a variety of devices and it worked as described. :+1: 

I would say if there any remaining style nits, fix them but otherwise :shipit:

**Pull request merged**